File: frontend/src/domain/knowledge.ts
export type ConceptType =
  | 'organization'
  | 'organization_unit'
  | 'person'
  | 'country'
  | 'region'
  | 'place'
  | 'technology'
  | 'equipment'
  | 'system'
  | 'project_program'
  | 'policy_law'
  | 'event'
  | 'document'

export type DocumentStatus = 'draft' | 'processing' | 'ready' | 'failed' | 'deleting'
export type EntityType = 'document' | 'chunk' | 'concept'

export type DocumentSummary = {
  id: number
  title: string
  filename: string | null
  source_type: 'paste' | 'upload'
  media_type: string
  summary: string
  keywords: string[]
  status: DocumentStatus
  character_count: number
  chunk_count: number
  concept_count: number
  created_at: string
  updated_at: string
}

export type DocumentChunk = {
  id: number
  document_id: number
  chunk_index: number
  content: string
  preview: string
  start_char: number
  end_char: number
  concept_ids: number[]
}

export type ConceptSummary = {
  id: number
  concept_type: ConceptType
  canonical_name: string
  english_name: string | null
  abbreviation: string | null
  description: string
  document_count: number
  chunk_count: number
  visibility: 'visible' | 'hidden' | 'orphaned'
}

export type ConceptAlias = {
  alias: string
  alias_type: 'ko' | 'en' | 'abbreviation' | 'source_mention'
  source_chunk_id: number | null
  confidence: number
}

export type ConceptRelation = {
  concept: ConceptSummary
  relation_type: string
  strength: number
  evidence_chunk_id: number | null
  explanation: string
}

export type ConceptDetail = ConceptSummary & {
  aliases: ConceptAlias[]
  source_chunks: DocumentChunk[]
  related_concepts: ConceptRelation[]
}

export type DocumentDetail = DocumentSummary & {
  chunks: DocumentChunk[]
  concepts: ConceptSummary[]
  original_content: string
  latest_job: AnalysisJob | null
}

export type GraphNode = {
  id: string
  entity_type: EntityType
  entity_id: number
  label: string
  subtype: ConceptType | null
  size: number
  color_token: string
  metadata: {
    title?: string
    summary?: string
    connection_count?: number
    status?: DocumentStatus
  }
}

export type GraphEdge = {
  id: string
  source: string
  target: string
  edge_type: 'contains' | 'mentions' | 'relates'
  relation_type: string | null
  strength: number
  is_directed: boolean
  evidence_chunk_id: number | null
}

export type GraphFilters = {
  conceptType: ConceptType | 'all'
  includeChunks: boolean
  recentDays: number | null
  minStrength: number
}

export type GraphSnapshot = {
  nodes: GraphNode[]
  edges: GraphEdge[]
  filters: GraphFilters
  truncated: boolean
  node_count: number
  edge_count: number
}

export type QuestionSource = {
  rank: number
  citation_key: string
  document_id: number | null
  chunk_id: number | null
  document_title: string
  document_status: 'ready' | 'deleted' | 'reanalyzed'
  chunk_preview: string
  start_char: number | null
  end_char: number | null
  score: number
  mapping_confidence: number
  openable: boolean
}

export type QuestionStatus = 'queued' | 'retrieving' | 'generating' | 'completed' | 'no_evidence' | 'failed'

export type QuestionResult = {
  id: number
  question: string
  status: QuestionStatus
  answer_markdown: string | null
  answer_language: string | null
  sources: QuestionSource[]
  related_concepts: ConceptSummary[]
  retrieval: {
    provider: 'vector_store' | 'lexical_fallback' | 'none'
    candidate_count: number
    returned_count: number
    mapping_failures: number
    top_score: number | null
    used_chunk_ids: number[]
  }
  created_at: string
  completed_at: string | null
}

export type QuestionHistorySummary = {
  id: number
  question_preview: string
  status: QuestionStatus
  answer_preview: string | null
  evidence_count: number
  created_at: string
  completed_at: string | null
}

export type AnalysisJob = {
  id: number
  document_id: number
  status: 'queued' | 'running' | 'completed' | 'failed' | 'cancel_requested' | 'canceled'
  stage: string
  progress: number
  message: string
  retry_count: number
  error: { code: string; message: string } | null
  started_at: string | null
  completed_at: string | null
}

export type AnalysisPreview = {
  kind: 'title' | 'summary' | 'keyword' | 'concept' | 'relation'
  payload: Record<string, string | number | null>
}

export type SystemStatus = {
  database: 'ready' | 'degraded'
  file_storage: 'ready' | 'degraded'
  openai_configured: boolean
  vector_store_configured: boolean
  analysis_running: number
  app_version: string
}

export const conceptTypeLabels: Record<ConceptType, string> = {
  organization: '조직',
  organization_unit: '조직 단위',
  person: '인물',
  country: '국가',
  region: '지역',
  place: '장소',
  technology: '기술',
  equipment: '장비',
  system: '체계',
  project_program: '사업·프로그램',
  policy_law: '정책·법률',
  event: '사건',
  document: '문서',
}

export const conceptTypeOrder = Object.keys(conceptTypeLabels) as ConceptType[]

File: backend/app/models/knowledge.py
from datetime import datetime, timezone

from sqlmodel import Field, SQLModel


def utc_now() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


class Document(SQLModel, table=True):
    __tablename__ = "documents"

    id: int | None = Field(default=None, primary_key=True)
    source_type: str = Field(index=True, max_length=20)
    original_filename: str | None = Field(default=None, max_length=255)
    media_type: str = Field(default="text/plain", max_length=120)
    storage_key: str = Field(max_length=500, unique=True)
    title: str = Field(max_length=255)
    summary: str = ""
    content_hash: str = Field(max_length=64, index=True)
    character_count: int = 0
    status: str = Field(default="draft", index=True, max_length=30)
    analysis_version: int = 0
    active_job_id: int | None = Field(default=None, index=True)
    vector_store_file_id: str | None = Field(default=None, index=True, max_length=120)
    vector_store_status: str = Field(default="not_uploaded", max_length=30)
    vector_store_error_code: str | None = Field(default=None, max_length=80)
    created_at: datetime = Field(default_factory=utc_now, index=True)
    updated_at: datetime = Field(default_factory=utc_now)
    deleted_at: datetime | None = None


class DocumentChunk(SQLModel, table=True):
    __tablename__ = "document_chunks"

    id: int | None = Field(default=None, primary_key=True)
    document_id: int = Field(foreign_key="documents.id", index=True)
    chunk_index: int = Field(index=True)
    start_char: int
    end_char: int
    content: str
    character_count: int
    content_hash: str = Field(max_length=64, index=True)
    created_at: datetime = Field(default_factory=utc_now)


class DocumentKeyword(SQLModel, table=True):
    __tablename__ = "document_keywords"

    document_id: int = Field(foreign_key="documents.id", primary_key=True)
    normalized_keyword: str = Field(primary_key=True, max_length=255)
    keyword: str = Field(max_length=255)
    rank: int = 0
    source: str = Field(default="ai", max_length=30)
    created_at: datetime = Field(default_factory=utc_now)


class Concept(SQLModel, table=True):
    __tablename__ = "concepts"

    id: int | None = Field(default=None, primary_key=True)
    concept_type: str = Field(index=True, max_length=40)
    canonical_name: str = Field(max_length=255)
    english_name: str | None = Field(default=None, max_length=255)
    abbreviation: str | None = Field(default=None, max_length=100)
    normalized_name: str = Field(index=True, max_length=255)
    description: str = Field(default="", max_length=500)
    merge_confidence: float = 1.0
    visibility: str = Field(default="visible", index=True, max_length=20)
    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)


class ConceptAlias(SQLModel, table=True):
    __tablename__ = "concept_aliases"

    id: int | None = Field(default=None, primary_key=True)
    concept_id: int = Field(foreign_key="concepts.id", index=True)
    alias: str = Field(max_length=255)
    normalized_alias: str = Field(index=True, max_length=255)
    alias_type: str = Field(max_length=30)
    source_chunk_id: int | None = Field(default=None, foreign_key="document_chunks.id")
    confidence: float = 1.0
    created_at: datetime = Field(default_factory=utc_now)


class ChunkConcept(SQLModel, table=True):
    __tablename__ = "chunk_concepts"

    chunk_id: int = Field(foreign_key="document_chunks.id", primary_key=True)
    concept_id: int = Field(foreign_key="concepts.id", primary_key=True)
    mention: str = Field(primary_key=True, max_length=255)
    mention_start: int | None = None
    mention_end: int | None = None
    extraction_confidence: float = 1.0
    description_snapshot: str = Field(default="", max_length=500)
    created_at: datetime = Field(default_factory=utc_now)


class ConceptRelation(SQLModel, table=True):
    __tablename__ = "concept_relations"

    id: int | None = Field(default=None, primary_key=True)
    source_concept_id: int = Field(foreign_key="concepts.id", index=True)
    target_concept_id: int = Field(foreign_key="concepts.id", index=True)
    relation_type: str = Field(max_length=80)
    is_directed: bool = True
    strength: float = 0.5
    extraction_confidence: float = 0.5
    explanation: str = Field(default="", max_length=500)
    evidence_chunk_id: int | None = Field(default=None, foreign_key="document_chunks.id", index=True)
    visibility: str = Field(default="visible", max_length=20)
    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)


class AnalysisJob(SQLModel, table=True):
    __tablename__ = "analysis_jobs"

    id: int | None = Field(default=None, primary_key=True)
    document_id: int = Field(foreign_key="documents.id", index=True)
    status: str = Field(default="queued", index=True, max_length=30)
    stage: str = Field(default="received", max_length=50)
    progress: int = 0
    message: str = ""
    analysis_version: int = 1
    retry_count: int = 0
    error_code: str | None = Field(default=None, max_length=80)
    error_message: str | None = None
    cancel_requested_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime = Field(default_factory=utc_now, index=True)
    updated_at: datetime = Field(default_factory=utc_now)


class QuestionHistory(SQLModel, table=True):
    __tablename__ = "question_histories"

    id: int | None = Field(default=None, primary_key=True)
    question: str
    status: str = Field(default="queued", index=True, max_length=30)
    answer_markdown: str | None = None
    answer_language: str | None = Field(default=None, max_length=10)
    model_name: str | None = Field(default=None, max_length=120)
    retrieval_count: int = 0
    citation_count: int = 0
    error_code: str | None = Field(default=None, max_length=80)
    error_message: str | None = None
    created_at: datetime = Field(default_factory=utc_now, index=True)
    completed_at: datetime | None = None


class QuestionSource(SQLModel, table=True):
    __tablename__ = "question_sources"

    id: int | None = Field(default=None, primary_key=True)
    question_history_id: int = Field(foreign_key="question_histories.id", index=True)
    rank: int
    chunk_id: int | None = Field(default=None, foreign_key="document_chunks.id")
    document_id: int | None = Field(default=None, foreign_key="documents.id")
    document_title_snapshot: str
    document_filename_snapshot: str | None = None
    chunk_content_snapshot: str
    start_char_snapshot: int | None = None
    end_char_snapshot: int | None = None
    score: float = 0.0
    citation_key: str
    mapping_confidence: float = 0.0
    current_state: str = "current"
    created_at: datetime = Field(default_factory=utc_now)


class AppSetting(SQLModel, table=True):
    __tablename__ = "app_settings"

    key: str = Field(primary_key=True, max_length=100)
    value: str
    updated_at: datetime = Field(default_factory=utc_now)
